File: src/swarm/core/pr_opened.py
"""REQ-71: structured GitHub PR-opened tool events (no markdown scrape).

A tool result becomes a ``pr_opened`` chrome event only when it already
carries a GitHub pull URL and/or explicit ``type: pr_opened``. Optional
branch / ``+N/-M`` / file counts are copied when present and never invented.
"""

from __future__ import annotations

import json
import re
from typing import Any

PR_OPENED_TYPE = "pr_opened"

# https://github.com/{owner}/{repo}/pull/{n} — no LAN, no :8001, no other hosts.
_GITHUB_PR_URL = re.compile(
    r"^https://github\.com/[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+/pull/\d+(?:/[A-Za-z0-9._~-]*)?(?:\?[^#]*)?(?:#.*)?$",
    re.IGNORECASE,
)

_URL_KEYS = ("url", "html_url", "pr_url", "pull_request_url")
_TITLE_KEYS = ("title", "name")
_NUMBER_KEYS = ("number", "pr_number", "pull_number")
_BRANCH_KEYS = ("branch", "head_ref", "head")
_ADDITIONS_KEYS = ("additions", "additions_count", "plus")
_DELETIONS_KEYS = ("deletions", "deletions_count", "minus")
_FILES_KEYS = ("files_changed", "changed_files", "files")
_STATUS_KEYS = ("status", "state")


def is_github_pr_url(value: object) -> bool:
    """True only for a public ``https://github.com/…/pull/N`` URL."""
    if not isinstance(value, str):
        return False
    text = value.strip()
    if not text or "://" not in text:
        return False
    lowered = text.lower()
    if lowered.startswith(("http://", "ws://", "wss://")):
        return False
    if "localhost" in lowered or "127.0.0.1" in lowered:
        return False
    if ":8001" in lowered:
        return False
    return bool(_GITHUB_PR_URL.match(text))


def _as_int(value: object) -> int | None:
    if isinstance(value, bool):
        return None
    if isinstance(value, int):
        return value
    if isinstance(value, float) and value.is_integer():
        return int(value)
    if isinstance(value, str) and value.strip().lstrip("+-").isdigit():
        try:
            return int(value.strip())
        except ValueError:
            return None
    return None


def _first_str(obj: dict[str, Any], keys: tuple[str, ...]) -> str | None:
    for key in keys:
        raw = obj.get(key)
        if isinstance(raw, str) and raw.strip():
            return raw.strip()
    return None


def _first_int(obj: dict[str, Any], keys: tuple[str, ...]) -> int | None:
    for key in keys:
        parsed = _as_int(obj.get(key))
        if parsed is not None:
            return parsed
    return None


def _github_pr_url_from(obj: dict[str, Any]) -> str | None:
    for key in _URL_KEYS:
        raw = obj.get(key)
        if is_github_pr_url(raw):
            return str(raw).strip()
    return None


def _opener_from(obj: dict[str, Any], *, agent_id: str = "", conversation_id: str = "") -> dict[str, str]:
    raw = obj.get("opener")
    opener: dict[str, str] = {}
    if isinstance(raw, dict):
        oid = raw.get("agent_id") or raw.get("agentId") or raw.get("id")
        name = raw.get("name")
        cid = raw.get("conversation_id") or raw.get("conversationId")
        if isinstance(oid, str) and oid.strip():
            opener["agent_id"] = oid.strip()
        if isinstance(name, str) and name.strip():
            opener["name"] = name.strip()
        if isinstance(cid, str) and cid.strip():
            opener["conversation_id"] = cid.strip()
    if agent_id and "agent_id" not in opener:
        opener["agent_id"] = agent_id
    if conversation_id and "conversation_id" not in opener:
        opener["conversation_id"] = conversation_id
    return opener


def _unwrap_candidate(value: Any) -> dict[str, Any] | None:
    if isinstance(value, dict):
        for nested_key in ("pull_request", "pr", "result", "data"):
            nested = value.get(nested_key)
            if isinstance(nested, dict) and (
                _github_pr_url_from(nested) or nested.get("type") == PR_OPENED_TYPE
            ):
                return nested
        return value
    return None


def parse_pr_opened(
    value: Any,
    *,
    agent_id: str = "",
    conversation_id: str = "",
) -> dict[str, Any] | None:
    """Return a chrome payload, or None when this is not a PR-opened result."""
    if isinstance(value, str):
        text = value.strip()
        if not text.startswith("{"):
            return None
        try:
            value = json.loads(text)
        except json.JSONDecodeError:
            return None
    obj = _unwrap_candidate(value)
    if obj is None:
        return None

    explicit = str(obj.get("type") or "").strip() == PR_OPENED_TYPE
    url = _github_pr_url_from(obj)
    number = _first_int(obj, _NUMBER_KEYS)
    title = _first_str(obj, _TITLE_KEYS)
    if not explicit and not url:
        return None
    if not explicit and url and number is None and title is None:
        # A lone GitHub link is not enough — require number or title.
        return None

    payload: dict[str, Any] = {"type": PR_OPENED_TYPE}
    if url:
        payload["url"] = url
    if number is not None:
        payload["number"] = number
    if title:
        payload["title"] = title

    branch = obj.get("branch")
    if not isinstance(branch, str) or not branch.strip():
        head = obj.get("head")
        if isinstance(head, dict):
            ref = head.get("ref")
            if isinstance(ref, str) and ref.strip():
                branch = ref
        elif isinstance(head, str) and head.strip() and "/" not in head.strip():
            branch = head
        else:
            branch = _first_str(obj, ("head_ref",))
    if isinstance(branch, str) and branch.strip():
        payload["branch"] = branch.strip()

    additions = _first_int(obj, _ADDITIONS_KEYS)
    if additions is not None:
        payload["additions"] = additions
    deletions = _first_int(obj, _DELETIONS_KEYS)
    if deletions is not None:
        payload["deletions"] = deletions
    files_changed = _first_int(obj, _FILES_KEYS)
    if files_changed is not None:
        payload["files_changed"] = files_changed

    status = _first_str(obj, _STATUS_KEYS)
    if status and status.lower() not in {PR_OPENED_TYPE, "tool_status", "tool_approval"}:
        payload["status"] = status

    opener = _opener_from(obj, agent_id=agent_id, conversation_id=conversation_id)
    if opener:
        payload["opener"] = opener
    return payload


def persist_pr_opened_message(messages: list[dict[str, Any]], payload: dict[str, Any]) -> None:
    """Append a status row so reload can rehydrate the card (not model context)."""
    content = json.dumps(payload, separators=(",", ":"))
    for row in messages:
        if row.get("role") == "status" and row.get("content") == content:
            return
    from swarm.core.transcript_roles import stamp_event

    messages.append(stamp_event({"role": "status", "kind": "status", "content": content}))

File: webui/frontend/src/lib/prOpened.ts
/**
 * REQ-71: structured GitHub PR-opened tool events (chrome, not markdown).
 * Missing optional fields stay omitted. Stats are never invented.
 */

export const PR_OPENED_TYPE = 'pr_opened' as const

export interface PrOpenedOpener {
  agentId: string
  name?: string
  conversationId?: string
}

export interface PrOpenedEvent {
  type: typeof PR_OPENED_TYPE
  url?: string
  number?: number
  title?: string
  branch?: string
  additions?: number
  deletions?: number
  filesChanged?: number
  status?: string
  opener?: PrOpenedOpener
}

const GITHUB_PR_URL =
  /^https:\/\/github\.com\/[A-Za-z0-9_.-]+\/[A-Za-z0-9_.-]+\/pull\/\d+(?:\/[A-Za-z0-9._~-]*)?(?:\?[^#]*)?(?:#.*)?$/i

export function isGithubPrUrl(value: unknown): value is string {
  if (typeof value !== 'string') return false
  const text = value.trim()
  if (!text) return false
  const lowered = text.toLowerCase()
  if (lowered.startsWith('http://') || lowered.startsWith('ws://')) return false
  if (lowered.includes('localhost') || lowered.includes('127.0.0.1')) return false
  if (lowered.includes(':8001')) return false
  return GITHUB_PR_URL.test(text)
}

function asInt(value: unknown): number | undefined {
  if (typeof value === 'boolean') return undefined
  if (typeof value === 'number' && Number.isFinite(value)) return Math.trunc(value)
  if (typeof value === 'string' && /^-?\d+$/.test(value.trim())) {
    return Number.parseInt(value.trim(), 10)
  }
  return undefined
}

function asTrimmed(value: unknown): string | undefined {
  if (typeof value !== 'string') return undefined
  const text = value.trim()
  return text || undefined
}

function pickUrl(obj: Record<string, unknown>): string | undefined {
  for (const key of ['url', 'html_url', 'pr_url', 'pull_request_url']) {
    const raw = obj[key]
    if (isGithubPrUrl(raw)) return raw.trim()
  }
  return undefined
}

function pickOpener(obj: Record<string, unknown>): PrOpenedOpener | undefined {
  const raw = obj.opener
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) return undefined
  const row = raw as Record<string, unknown>
  const agentId = asTrimmed(row.agentId ?? row.agent_id ?? row.id)
  if (!agentId) return undefined
  const opener: PrOpenedOpener = { agentId }
  const name = asTrimmed(row.name)
  if (name) opener.name = name
  const conversationId = asTrimmed(row.conversationId ?? row.conversation_id)
  if (conversationId) opener.conversationId = conversationId
  return opener
}

function unwrapCandidate(value: unknown): Record<string, unknown> | null {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return null
  const obj = value as Record<string, unknown>
  for (const key of ['pull_request', 'pr', 'result', 'data']) {
    const nested = obj[key]
    if (nested && typeof nested === 'object' && !Array.isArray(nested)) {
      const inner = nested as Record<string, unknown>
      if (inner.type === PR_OPENED_TYPE || pickUrl(inner)) return inner
    }
  }
  return obj
}

export function parsePrOpened(value: unknown): PrOpenedEvent | null {
  let raw: unknown = value
  if (typeof raw === 'string') {
    const text = raw.trim()
    if (!text.startsWith('{')) return null
    try {
      raw = JSON.parse(text) as unknown
    } catch {
      return null
    }
  }
  const obj = unwrapCandidate(raw)
  if (!obj) return null

  const explicit = String(obj.type || '').trim() === PR_OPENED_TYPE
  const url = pickUrl(obj)
  const number = asInt(obj.number ?? obj.pr_number ?? obj.pull_number)
  const title = asTrimmed(obj.title ?? obj.name)
  if (!explicit && !url) return null
  if (!explicit && url && number === undefined && !title) return null

  const event: PrOpenedEvent = { type: PR_OPENED_TYPE }
  if (url) event.url = url
  if (number !== undefined) event.number = number
  if (title) event.title = title

  let branch = asTrimmed(obj.branch)
  if (!branch) {
    const head = obj.head
    if (head && typeof head === 'object' && !Array.isArray(head)) {
      branch = asTrimmed((head as Record<string, unknown>).ref)
    } else if (typeof head === 'string' && head.trim() && !head.includes('/')) {
      branch = head.trim()
    } else {
      branch = asTrimmed(obj.head_ref)
    }
  }
  if (branch) event.branch = branch

  const additions = asInt(obj.additions ?? obj.additions_count ?? obj.plus)
  if (additions !== undefined) event.additions = additions
  const deletions = asInt(obj.deletions ?? obj.deletions_count ?? obj.minus)
  if (deletions !== undefined) event.deletions = deletions
  const filesChanged = asInt(obj.files_changed ?? obj.changed_files ?? obj.files)
  if (filesChanged !== undefined) event.filesChanged = filesChanged

  const status = asTrimmed(obj.status ?? obj.state)
  if (status && status !== PR_OPENED_TYPE && status !== 'tool_status') {
    event.status = status
  }

  const opener = pickOpener(obj)
  if (opener) event.opener = opener
  return event
}

export function formatPrFileStats(event: PrOpenedEvent): string | undefined {
  const parts: string[] = []
  if (typeof event.additions === 'number') parts.push(`+${event.additions}`)
  if (typeof event.deletions === 'number') parts.push(`-${event.deletions}`)
  return parts.length ? parts.join(' ') : undefined
}

function normId(value: string | undefined): string {
  return (value || '').trim().toLowerCase()
}

/** True when the card is already on the opener's agent + thread. */
export function isSameOpenerChat(
  opener: PrOpenedOpener | undefined,
  current: { agentId?: string; conversationId?: string },
): boolean {
  if (!opener?.agentId) return true
  if (normId(opener.agentId) !== normId(current.agentId)) return false
  if (opener.conversationId && normId(opener.conversationId) !== normId(current.conversationId)) {
    return false
  }
  return true
}

export function openerChatSearch(opener: PrOpenedOpener): URLSearchParams {
  const params = new URLSearchParams()
  params.set('blueprint', opener.agentId)
  if (opener.conversationId) params.set('session', opener.conversationId)
  return params
}
